### PR TITLE
[Merged by Bors] - Makefile cleanup

### LINKED ISSUE #1164: Refactor Makefile recipes
Our Makefile has gotten pretty unweildy lately and it needs some love. Here are some things we want to improve about it:

- Make sure recipes are organized so that we can distinguish between compiling and testing
  - We should have recipes to "build _and_ test" but also recipes to "test only without building". This will help with caching in CI
- Ensure we can distinguish between recipes that build "client" dependencies (e.g. the `fluvio` binary) and those that build "server" dependencies (e.g. the `fluvio-run` binary)
- Come up with consistent recipe naming schemes (e.g. use either `-` or `_` but not both)
- Deduplicate different configurations of the same test, e.g. `smoke-test-*`

### PR DESCRIPTION
**Pull request by nicholastmosher**
First part of #1164.

- Splits up smoke-test recipes such that they may be run with or without building

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -17,14 +17,29 @@ CLIENT_LOG=warn
 SERVER_LOG=fluvio=debug
 TEST_BIN=$(if $(TARGET),./target/$(TARGET)/$(BUILD_PROFILE)/flv-test,./target/$(BUILD_PROFILE)/flv-test)
 TEST_LOG=--client-log ${CLIENT_LOG} --server-log ${SERVER_LOG}
-DEFAULT_SPU=2
```
Comment (sehz) [inline Makefile:20]: should keep this.  it's simple way to configure SPU

File: Makefile [hunk 2]
```diff
@@ -17,14 +17,29 @@ CLIENT_LOG=warn
...
+TEST_ARG_REPLICATION=-r 2
+TEST_ARG_LOCAL=
+TEST_ARG_DEVELOP=
+TEST_ARG_TLS=
+TEST_ARG_KEEP_CLUSTER=
+TEST_ARG_AUTH_CONFIG_MAP=
+TEST_ARG_SKIP_CHECKS=
+TEST_ARG_EXTRA=
+TEST_ARG_CONSUMER_WAIT=
+TEST_ARG_PRODUCER_ITERATION=--producer-iteration=1000
```
Comment (sehz) [inline Makefile:41]: should use DEFAULT_ITERATION

File: Makefile [hunk 3]
```diff
@@ -52,22 +67,53 @@ endif
...
-
-smoke-test-tls:	test-clean-up build_test
-	$(TEST_BIN) smoke --spu ${DEFAULT_SPU} --tls --local ${TEST_LOG} ${SKIP_CHECK} -- --producer-iteration=${DEFAULT_ITERATION}
-
-smoke-test-tls-policy:	test-clean-up build_test
-	AUTH_POLICY=$(SC_AUTH_CONFIG)/policy.json X509_AUTH_SCOPES=$(SC_AUTH_CONFIG)/scopes.json  \
-	FLV_SPU_DELAY=$(SPU_DELAY) \
-	$(TEST_BIN) smoke --spu ${DEFAULT_SPU} --tls --local ${TEST_LOG} ${SKIP_CHECK} --keep-cluster -- --producer-iteration=${DEFAULT_ITERATION}
+smoke-test-base: test-setup smoke-test-base-nobuild
+smoke-test-base-nobuild: test-setup-nobuild
```
Comment (sehz) [inline Makefile:71]: can we shorten this target name?

File: Makefile [hunk 4]
```diff
@@ -52,22 +67,53 @@ endif
...
+			${TEST_ARG_REPLICATION} \
+			${TEST_ARG_KEEP_CLUSTER} \
+			${TEST_ARG_AUTH_CONFIG_MAP} \
+			${TEST_ARG_SKIP_CHECKS} \
+			${TEST_ARG_EXTRA} \
+			-- \
+			${TEST_ARG_CONSUMER_WAIT} \
+			${TEST_ARG_PRODUCER_ITERATION}
+
+smoke-test-nobuild: TEST_ARG_LOCAL=--local
```
Comment (sehz) [inline Makefile:90]: rename target as `smoke-test-local`

File: Makefile [hunk 5]
```diff
@@ -52,22 +67,53 @@ endif
...
+smoke-test-nobuild: TEST_ARG_SKIP_CHECKS=--skip-checks
+smoke-test-nobuild: smoke-test-base-nobuild
+smoke-test: test-setup smoke-test-nobuild
+
+smoke-test-stream-nobuild: TEST_ARG_SKIP_CHECKS+=--skip-checks
```
Comment (sehz) [inline Makefile:95]: ditto.  `smoke-test-stream`

File: Makefile [hunk 6]
```diff
@@ -52,22 +67,53 @@ endif
...
+smoke-test-stream-nobuild: TEST_ARG_CONSUMER_WAIT+=--consumer-wait=true
+smoke-test-stream-nobuild: smoke-test-base-nobuild
+smoke-test-stream: test-setup smoke-test-stream-nobuild
+
+smoke-test-tls-nobuild: TEST_ARG_TLS=--tls
```
Comment (sehz) [inline Makefile:100]: smoke-test-local-tls

File: Makefile [hunk 7]
```diff
@@ -52,22 +67,53 @@ endif
...
+smoke-test-tls-nobuild: TEST_ARG_LOCAL=--local
+smoke-test-tls-nobuild: smoke-test-base-nobuild
+smoke-test-tls: test-setup smoke-test-tls-nobuild
+
+smoke-test-tls-policy-nobuild: TEST_ENV_AUTH_POLICY=$(SC_AUTH_CONFIG)/policy.json X509_AUTH_SCOPES=$(SC_AUTH_CONFIG)/scopes.json
```
Comment (sehz) [inline Makefile:105]: use same target name as before

**Review by sehz — CHANGES_REQUESTED** [review]

File: Makefile (hunk 5 above)
Comment (nicholastmosher) [reply Makefile:95]: Did you see that each recipe has two variants? There is `smoke-test-stream-nobuild` and `smoke-test-stream`, which _will_ build. It's expected that developers will mostly use `smoke-test-stream` and CI will use `smoke-test-stream-nobuild`

File: Makefile (hunk 3 above)
Comment (nicholastmosher) [reply Makefile:71]: Do you mean shorten `smoke-test-base` or `smoke-test-base-nobuild`?

File: Makefile (hunk 7 above)
Comment (nicholastmosher) [reply Makefile:105]: There are both `smoke-test-tls-policy-nobuild` and `smoke-test-tls-policy`

File: Makefile (hunk 6 above)
Comment (nicholastmosher) [reply Makefile:100]: Note there are both `smoke-test-tls-nobuild` and also `smoke-test-tls`

File: Makefile [hunk 8]
```diff
@@ -17,14 +17,29 @@ CLIENT_LOG=warn
...
+# Test env
+TEST_ENV_AUTH_POLICY=
+TEST_ENV_FLV_SPU_DELAY=
+
+# Test args
+TEST_ARG_SPU=--spu 2
+TEST_ARG_LOG=--client-log ${CLIENT_LOG} --server-log ${SERVER_LOG}
+TEST_ARG_REPLICATION=-r 2
+TEST_ARG_LOCAL=
+TEST_ARG_DEVELOP=
```
[the remark this answers is not in the document]
Comment (sehz) [reply Makefile:36]: let's try to keep existing frequently used param.  everything else can be moved to EXTRA_ARG

File: Makefile (hunk 3 above)
Comment (sehz) [reply Makefile:73]: smoke-test-base-nobuild

File: Makefile (hunk 6 above)
Comment (sehz) [reply Makefile:102]: keep only used by CI

File: Makefile (hunk 3 above)
Comment (nicholastmosher) [reply Makefile]: I changed `smoke-test` to `smoke-test-local` and I changed `smoke-test-base` to `smoke-test`

**Comment by nicholastmosher** [comment]
@sehz how does this look?

**Review by sehz — CHANGES_REQUESTED** [review]
minor nits. also add to CHANGELOG

File: Makefile [hunk 9]
```diff
@@ -52,22 +64,36 @@ endif
...
-
-smoke-test-tls-policy:	test-clean-up build_test
-	AUTH_POLICY=$(SC_AUTH_CONFIG)/policy.json X509_AUTH_SCOPES=$(SC_AUTH_CONFIG)/scopes.json  \
-	FLV_SPU_DELAY=$(SPU_DELAY) \
-	$(TEST_BIN) smoke --spu ${DEFAULT_SPU} --tls --local ${TEST_LOG} ${SKIP_CHECK} --keep-cluster -- --producer-iteration=${DEFAULT_ITERATION}
+smoke-test: test-setup
+	# Set ENV
+	AUTH_POLICY=$(TEST_ENV_AUTH_POLICY) \
+	FLV_SPU_DELAY=$(TEST_ENV_FLV_SPU_DELAY) \
+		$(TEST_BIN) smoke \
```
[the remark this answers is not in the document]
Comment (nicholastmosher) [reply Makefile:72]: I did this to distinguish between the environment variables (above) from the command invocation and args (below)

**Review by sehz — APPROVED** [review]

**Pull request closed without merging**